dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/DLCOracle.scala

```diff
@@ -91,85 +91,207 @@ case class DLCOracle(private val extPrivateKey: ExtPrivateKeyHardened)(implicit
+
+    val descriptorTLV = UninitializedEnumEventDescriptorV0(outcomes)
+
+    createNewEvent(eventName, maturationTime, descriptorTLV)
+  }
+
+  def createNewEvent(
+      eventName: String,
+      maturationTime: Instant,
+      uninitializedDescriptor: UninitializedEventDescriptor[
+        EventDescriptorTLV]): Future[OracleAnnouncementTLV] = {
+    require(maturationTime.isAfter(TimeUtil.now),
+            s"Event cannot mature in the past, got $maturationTime")
+
     for {
       indexOpt <- rValueDAO.maxKeyIndex
-      index = indexOpt match {
+      firstIndex = indexOpt match {
         case Some(value) => value + 1
         case None        => 0
       }
 
       signingVersion = SigningVersion.latest
 
-      path = getPath(index)
-      nonce = getKValue(eventName, path, signingVersion).schnorrNonce
-
-      hash = CryptoUtil.taggedSha256(
-        nonce.bytes ++ CryptoUtil.serializeForHash(eventName) ++ ByteVector
-          .fromLong(maturationTime.getEpochSecond),
-        signingVersion.announcementTag)
-      announcementSignature = signingKey.schnorrSign(hash.bytes)
-
-      rValueDb = RValueDbHelper(nonce = nonce,
-                                eventName = eventName,
-                                account = rValAccount,
-                                chainType = rValueChainIndex,
-                                keyIndex = index,
-                                announcementSignature = announcementSignature)
-
-      eventDb = EventDb(nonce,
-                        publicKey,
-                        eventName,
-                        outcomes.size,
-                        signingVersion,
-                        maturationTime,
-                        None)
-      eventOutcomeDbs = outcomes.map { outcome =>
-        val hash = CryptoUtil.taggedSha256(outcome, signingVersion.outcomeTag)
-        EventOutcomeDb(nonce, outcome, hash)
+      rValueDbs =
+        0.until(uninitializedDescriptor.noncesNeeded)
+          .map { num =>
+            val index = firstIndex + num
+            val path = getPath(index)
+            val nonce = getKValue(eventName, path, signingVersion).schnorrNonce
+
+            RValueDbHelper(nonce = nonce,
+                           eventName = eventName,
+                           account = rValAccount,
+                           chainType = rValueChainIndex,
+                           keyIndex = index)
+          }
+          .toVector
+
+      descriptorTLV = uninitializedDescriptor.addNonces(rValueDbs.map(_.nonce))
+
+      epoch = UInt32(maturationTime.getEpochSecond)
+
+      eventTLV = OracleEventV0TLV(publicKey, epoch, descriptorTLV, eventName)
+
+      announcementBytes = signingVersion.calcAnnouncementBytes(eventTLV)
+      announcementSignature = signingKey.schnorrSign(announcementBytes)
+
+      eventOutcomeDbs = descriptorTLV match {
+        case enum: EnumEventDescriptorV0TLV =>
+          require(rValueDbs.size == 1)
```

Comment:
I guess this invariant is obvious to me, can you add a error message so readers can understand what this invariant is required? Plus this gives useful error messages for users of the API.